### DISCUSSION TITLE
多层pag叠加问题求助

### DISCUSSION THREAD

**Discussion by zhuiyi007** [Tencent/libpag#2334] [General] [answered]
我想要做一个两层pag资源叠加的效果，背景层需要对原素材做一些挖孔透出，前景层正常播放，目前我是使用了两层PAGView做的，在背景层的view上增加一层layer做的遮罩裁切效果，效果可以符合预期，但是问题是两层PAGView同时播放会存在内存消耗较大的问题，担心会在一些低端机上无法加载的问题。
我想在native侧针对背景层的PAG资源先做一次加工，事先将孔挖好，然后再将两层合为一层，再进行播放。由于挖孔的尺寸，位置需要根据不同机型来进行适配，所以不能通过事先在PAG资源中预留位置的方式进行替换（或者说是否可以有一些动态参数可以支持替换资源位置的动态性）。
这么做的原因是在本地通过demo进行简单测试后发现，合成一个PAG资源后播放一个PAGView比播放两个PAGView在内存方面会降低50M左右。
想问下有没有方式可以实现这个操作？

**Comment by kevingpqi123** [comment]
听你的描述实现的有些复杂，建议描述清楚你要实现的功能以及效果图，我们可以推荐一些解决方案，pag 已经发展了这么多年，你遇到的问题我们之前基本上都遇到过

**Reply by zhuiyi007** [reply]
我想实现的效果大概是这样的：
<img width="416" alt="image" src="https://github.com/Tencent/libpag/assets/10155576/2b45ae70-0b39-4e64-afa1-d8e07dff614f">
这个效果是由两层pag组成，具体的层级结构如图：
<img width="1388" alt="image" src="https://github.com/Tencent/libpag/assets/10155576/1a75615c-8b14-4780-84b0-bb9e19b9130a">
用这种方式实现一种素材“出框”的效果。
因为播放两个pag素材比播放一个pag素材要耗性能，所以我们想在业务层将背景pag按照绿色方框给抠好，再跟前景合成一个新的pag（前景的透明效果是在原始素材中就已经抠好了），然后使用一层pagView去播放。
但是发现好像没有相关接口能够在客户端上对pag素材进行相关的裁剪。

**Reply by kevingpqi123** [reply]
PAGLayer 有一个 setMatrix 的接口，可以对 PAGLayer 、PAGComposition、PAGFile 进行各种矩阵变化

**Reply by zhuiyi007** [reply]
这个接口我有看过，他只能对原素材做整体的一些形变，但是我只是想截取原素材内的某些局部范围内的素材，貌似不能满足需求

**Reply by kevingpqi123** [reply]
那你可以直接获取整体渲染结果，然后进行裁剪获取局部的

**Reply by domchen** [reply, marked as answer]
> 这个接口我有看过，他只能对原素材做整体的一些形变，但是我只是想截取原素材内的某些局部范围内的素材，貌似不能满足需求

创建一个空的PAGComposition，设置好要裁剪的宽高就行。然后把你要裁剪的PAGFile添加进去，通过设置setMatrix接口挪动位置，超过父级PAGComposition宽高的位置都不会显示。组合就是这样用的。
